cmd/xl-storage.go

```diff
@@ -1140,7 +1141,17 @@ func (s *xlStorage) deleteVersions(ctx context.Context, volume, path string, fis
 		return err
 	}
 
+	latestV, err := xlMeta.getIdx(0)
+	if err != nil {
+		return err
+	}
```

Comment:
No need for a full unmarshal. You can add this helper, when simply do `if xlMeta.latestType() == DeleteType`.

```
// latestType returns the latest version type.
// returns invalidVersionType if no versions exist.
func (x *xlMetaV2) latestType() VersionType {
	if x == nil || len(x.versions) == 0 {
		return invalidVersionType
	}
	return x.versions[0].header.Type
}
```

Reply:
Also doing it in the loop will keep up with mutations (ie as deletes are happening).

A bit worried that the top delete marker could end up getting deleted if the fis contain the delete marker version id.

Couldn't we then end up with an object that *isn't* deleted?

Somehow it feels like we should apply all version-id deletes, and then after that see if we still should add the delete marker... Annoying.. Maybe you can find another way?

Reply:
re-ping @vadmeste ^^

Reply:
> Also doing it in the loop will keep up with mutations (ie as deletes are happening).

I am not sure I understand, we have object locks already, are you talking about something else ?

> A bit worried that the top delete marker could end up getting deleted if the fis contain the delete marker version id.

FileInfo.Deleted member  is false when you want to remove a delete marker and you know its version-id

Reply:
@vadmeste Sure.